subsys/bluetooth/host/l2cap.c

```diff
@@ -1310,6 +1312,13 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	if (ecred_cb && ecred_cb->ecred_conn_req) {
 		ecred_cb->ecred_conn_req(conn, result, psm);
 	}
+
+	for (i = 0; i < req_cid_count; i++) {
+		/* Raise connected callback for established channels */
+		if ((dcid[i] != 0x00) && (chan[i]->ops->connected != NULL)) {
+			chan[i]->ops->connected(chan[i]);
+		}
+	}
```

Comment:
I think the connected callback should not fire until/unless a conn_rsp PDU with BT_L2CAP_LE_SUCCESS has been queued on the connection TX queue.

```suggestion
	if (rsp_queued && result == BT_L2CAP_LE_SUCCESS) {
		for (i = 0; i < req_cid_count; i++) {
			/* Raise connected callback for established channels */
			if ((dcid[i] != 0x00) && (chan[i]->ops->connected != NULL)) {
				chan[i]->ops->connected(chan[i]);
			}
		}
	}
```

`rsp_queued` is not defined yet. I suggest manually inlining `l2cap_send` and testing the return code of `bt_l2cap_send`. A similar inlining should be done in `le_conn_req`.

Reply:
Makes sense @alwa-nordic, thanks for spotting this - I'll amend the PR